textproc/qpdf/Portfile

```diff
@@ -20,5 +20,5 @@ long_description    QPDF is a command-line program that does structural, \
 homepage            http://qpdf.sourceforge.net/
 github.tarball_from releases
 
-depends_lib         port:jpeg \
+depends_lib         lib:libjpeg:libjpeg-turbo \
```

Comment:
Two problems here:
 * the dependency will be satisfied by a libjpeg in /usr/lib. You could use a path: dependency instead to avoid this, but...
 * libjpeg-turbo is not ABI compatible with jpeg.

Reply:
As per <https://trac.macports.org/ticket/38907#comment:15>

How might we resolve this issue for folks... it seems a big enough headache to justify a proper solution. Like the +x11/+quartz issue? Or is that just too much headache for something like a jpeg library abi issue, and we just tell people to stuff it, MacPorts uses port:jpeg, and be done with it?

Reply:
As I've said before, I'd be fine with switching everything to libjpeg-turbo if someone wants to do the work.

Reply:
> * the dependency will be satisfied by a libjpeg in /usr/lib. You could use a path: dependency instead to avoid this, but...

Indeed. I didn't read the documentation closely enough. PR updated to use `path`.

> * libjpeg-turbo is not ABI compatible with jpeg.

libjpeg-turbo is ABI-compatible with libjpeg v8 but not with new interfaces added to libjpeg v9. qpdf is explicitly documented as using only the v8 ABI and being compatible with libjpeg-turbo: https://github.com/qpdf/qpdf/blob/master/README.md#prerequisites

Reply:
libjpeg-turbo is ABI-compatible with libjpeg v8 but not with new interfaces added to libjpeg v9. qpdf is explicitly documented as using only the v8 ABI and being compatible with libjpeg-turbo: https://github.com/qpdf/qpdf/blob/master/README.md#prerequisites

Reply:
Yes, but the jpeg port is version 9. A binary will be linked with either libjpeg.9.dylib or libjpeg.8.dylib and not having that version installed will make it fail to load.

Reply:
True, but MacPorts handles this gracefully. If libjpeg is removed and replaced with libjpeg-turbo (or vice versa), MacPorts marks qpdf as broken and asks to rebuild it against the new library):

```
# port uninstall jpeg         
Note: It is not recommended to uninstall/deactivate a port that has dependents as it breaks the dependents.
The following ports will break: qpdf @9.0.0_0
Continue? [y/N]: y
Warning: Uninstall forced.  Proceeding despite dependencies.
--->  Deactivating jpeg @9c_0
--->  Cleaning jpeg
--->  Uninstalling jpeg @9c_0
--->  Cleaning jpeg
# port install libjpeg-turbo
--->  Computing dependencies for libjpeg-turbo
[...]
--->  Cleaning libjpeg-turbo
--->  Scanning binaries for linking errors
--->  Found 3 broken files, matching files to ports
--->  Found 1 broken port, determining rebuild order
You can always run 'port rev-upgrade' again to fix errors.
The following ports will be rebuilt: qpdf @9.0.0
Continue? [Y/n]: 
```

Reply:
@jmroot: I believe that the incompatibility between library versions is already adequately addressed by MacPorts's behaviour around automatically rebuilding a "broken" port. I also observe that this same pattern (allowing either libjpeg or libjpeg-turbo to be installed via a `path` dependency) is already used by four other ports. On these grounds, are you prepared to mark this as accepted?

Reply:
The **current** MacPorts expectation is that substitute libraries like this that provide similar functionality need to be drop-in no-rebuild replacements, ie people who get them off the buildbot would not need to rebuild. There are several libraries like that in MacPorts.

Personally, I'm leaning towards what you are asking for, which is the expectation be changed so that people like you can use a similar but not ABI-identical library. But the long-standing MacPorts admins find this approach is (at present) not an acceptable way to go.

So as far as I can see, unless that position changes, you are out of luck with this kind of PR as it will not be accepted.

You should know that you can have your own personal MacPorts overlay repo and run it however you darned well want to run it for your own builds and your own installation. It is easy, even trivial, to do that, and I have such a repo for virtually every single MacPorts installation I have, for exactly these kinds of reasons.

Reply:
@kencu: Thanks for explaining the background context. As long as this is the policy I guess there's no way to make this option available. Providing libjpeg-turbo support as a variant isn't really feasible, as it would basically preclude installing any other ports that depend on libjpeg.